cpp/src/join/conditional_join.cu

```diff
@@ -13,31 +13,234 @@
+  }
+  return join_indices;
+}
+
+std::size_t compute_conditional_join_output_size(table_view const& left,
+                                                 table_view const& right,
+                                                 ast::expression binary_predicate,
+                                                 null_equality compare_nulls,
+                                                 join_kind JoinKind,
+                                                 rmm::cuda_stream_view stream,
+                                                 rmm::mr::device_memory_resource* mr)
+{
+  // We can immediately filter out cases where the right table is empty. In
+  // some cases, we return all the rows of the left table with a corresponding
+  // null index for the right table; in others, we return an empty output.
+  if (right.num_rows() == 0) {
+    switch (JoinKind) {
+      // Left, left anti, and full (which are effectively left because we are
+      // guaranteed that left has more rows than right) all return a all the
+      // row indices from left with a corresponding NULL from the right.
+      case join_kind::LEFT_JOIN:
+      case join_kind::LEFT_ANTI_JOIN:
+      case join_kind::FULL_JOIN: return left.num_rows();
+      // Inner and left semi joins return empty output because no matches can exist.
+      case join_kind::INNER_JOIN:
+      case join_kind::LEFT_SEMI_JOIN: return 0;
+    }
+  }
+
+  // Prepare output column. Whether or not the output column is nullable is
+  // determined by whether any of the columns in the input table are nullable.
+  // If none of the input columns actually contain nulls, we can still use the
+  // non-nullable version of the expression evaluation code path for
+  // performance, so we capture that information as well.
+  auto const nullable  = cudf::nullable(left) || cudf::nullable(right);
+  auto const has_nulls = nullable && (cudf::has_nulls(left) || cudf::has_nulls(right));
+
+  auto const parser =
+    ast::detail::expression_parser{binary_predicate, left, right, has_nulls, stream, mr};
+  CUDF_EXPECTS(parser.output_type().id() == type_id::BOOL8,
+               "The expression must produce a boolean output.");
+
+  auto left_table  = table_device_view::create(left, stream);
+  auto right_table = table_device_view::create(right, stream);
+
+  // Allocate storage for the counter used to get the size of the join output
+  rmm::device_scalar<std::size_t> size(0, stream, mr);
+  CHECK_CUDA(stream.value());
+  detail::grid_1d config(left_table->num_rows(), DEFAULT_JOIN_BLOCK_SIZE);
+  auto const shmem_size_per_block =
+    parser.device_expression_data.shmem_per_thread * config.num_threads_per_block;
+
+  // Determine number of output rows without actually building the output to simply
+  // find what the size of the output will be.
+  join_kind KernelJoinKind = JoinKind == join_kind::FULL_JOIN ? join_kind::LEFT_JOIN : JoinKind;
```

Comment:
I'm working on the Java bindings for this, and the full join unit tests are failing.  I suspect the problem is here. 
 Treating a full join as a left join will not produce the correct row count for a full join.

Reply:
You're right, sorry about that. This was careless on my part. The full join requires an additional kernel launch to find all rows in the right table that are not present in the left table because the simplest and most efficient thread configuration was a 1D grid over the left table rows. We basically just do a left join and then add the missing right table rows in a separate launch, and as a result, the full join size is initially just computed as the left join size. I should be able to repurpose the logic in `hash_join::full_join_size` and/or `detail::get_left_join_indices_complement` to correct this error.

Reply:
Discussed offline: Both the hash join and conditional join APIs for computing full join size would require manifesting the left join indices in order to find out which rows in the right table had no matches in the left table. The size calculation therefore isn't particularly useful for full joins since the purpose is to provide a means to detect the potential for overflow or OOM errors and perform a distributed calculation. As a result, I'm going to disable full join size calculation in this PR for now and remove the corresponding hash join API as well. If we find a non-Spark use case in the future, we can reassess the tradeoffs between an API that manifests the left join in order to compute an accurate full join size, or an API that estimates the full join size as the size of a left join plus the number of rows in the right table (which would be an upper bound).